compiler/AST/foralls.cpp

```diff
@@ -230,3 +235,495 @@ bool astUnderFI(const Expr* ast, ForallIntents* fi) {
+
+  if (fast) {
+
+    if (zippered) {
+      followBlock->insertAtTail("'move'(%S, _getIteratorZip(_toFastFollowerZip(%S, %S)))", followIter, iter, leadIdxCopy);
+    } else {
+      followBlock->insertAtTail("'move'(%S, _getIterator(_toFastFollower(%S, %S)))",       followIter, iter, leadIdxCopy);
+    }
+  } else {
+
+    if (zippered) {
+      followBlock->insertAtTail("'move'(%S, _getIteratorZip(_toFollowerZip(%S, %S)))",     followIter, iter, leadIdxCopy);
+    } else {
+      followBlock->insertAtTail("'move'(%S, _getIterator(_toFollower(%S, %S)))",           followIter, iter, leadIdxCopy);
+    }
+  }
+
+  ref->insertAfter(followBlock); // otherwise it wouldn't normalize
+  normalize(followBlock);
+  followBlock->remove();
+
+  followBlock->insertAtTail(new DefExpr(followIdx));
+
+  followBlock->insertAtTail("{TYPE 'move'(%S, iteratorIndex(%S)) }", followIdx, followIter);
+
+  followBlock->insertAtTail(followBody);
+  followBlock->insertAtTail(new CallExpr("_freeIterator", followIter));
+
+  return followBlock;
+}
+
+// Return NULL if the def is not found or is uncertain.
+static Expr* findDefOfTemp(SymExpr* origSE)
+{
+  Symbol* origSym = origSE->symbol();
+  if (!origSym->hasFlag(FLAG_TEMP)) return NULL;  // only temps
+
+  SymExpr* otherSE = origSym->getSingleDef();
+
+  if (CallExpr* def = toCallExpr(otherSE->parentExpr))
+    if (def->isPrimitive(PRIM_MOVE))
+      if (otherSE == def->get(1))
+        return def->get(2);
+
+  // uncertain situation
+  return NULL;
+}
+
+static void removeOrigIterCall(SymExpr* origSE)
+{
+  INT_ASSERT(!origSE->inTree());
+
+  Symbol* origSym = origSE->symbol();
+  INT_ASSERT(origSym->hasFlag(FLAG_TEMP));
+
+  SymExpr* otherSE = origSym->firstSymExpr();
+  // Only one SE for this symbol
+  INT_ASSERT(origSym->lastSymExpr() == otherSE);
+  
+  if (CallExpr* def = toCallExpr(otherSE->parentExpr))
+    if (def->isPrimitive(PRIM_MOVE))
+      if (otherSE == def->get(1)) {
+        // Remove origSym - except caller will remove origSE.
+        def->remove();
+        origSym->defPoint->remove();
+        return;
+      }
+
+  // The above did not succeed.
+  INT_ASSERT(false);
+}
+
+// Replaces 'origSE' in the tree with the result.
+static CallExpr* buildForallParIterCall(ForallStmt* pfs, SymExpr* origSE)
+{
+  CallExpr* iterCall = NULL;
+
+  if (origSE->symbol()->type->symbol->hasFlag(FLAG_ITERATOR_RECORD)) {
+    // Our iterable expression is an iterator call.
+
+    CallExpr* origIterCall = toCallExpr(findDefOfTemp(origSE));
+    // What to do if we do not find it?
+    INT_ASSERT(origIterCall);
+
+    FnSymbol* origTarget = origIterCall->resolvedFunction();
+    INT_ASSERT(origTarget);
+
+    const char* targetName = origTarget->name;
+    if (!strncmp(targetName, "_parloopexpr", 12))
+      // a forall loop over a (possibly zippered) forall expression, ex.:
+      //  test/reductions/deitz/test_maxloc_reduce_wmikanik_bug2.chpl
+      targetName = astr("_iterator_for_loopexpr", targetName + 12);
```

Comment:
Am I understanding that this is an artifact of not using ForallStmt everywhere we could?